File: config/src/artifact/vorpal_shell.rs
use crate::artifact::SYSTEMS;
use anyhow::Result;
use vorpal_sdk::artifact::{
    bun::Bun,
    crane::Crane,
    go::Go,
    goimports::Goimports,
    gopls::Gopls,
    grpcurl::Grpcurl,
    language::go::{get_goarch, get_goos},
    nodejs::NodeJS,
    pnpm::Pnpm,
    protoc::Protoc,
    protoc_gen_go::ProtocGenGo,
    protoc_gen_go_grpc::ProtocGenGoGrpc,
    staticcheck::Staticcheck,
};
use vorpal_sdk::{artifact, context::ConfigContext};

#[derive(Default)]
pub struct VorpalShell {}

impl VorpalShell {
    pub fn new() -> Self {
        Self::default()
    }

    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let bun = Bun::new().build(context).await?;
        let crane = Crane::new().build(context).await?;
        let go = Go::new().build(context).await?;
        let goimports = Goimports::new().build(context).await?;
        let gopls = Gopls::new().build(context).await?;
        let grpcurl = Grpcurl::new().build(context).await?;
        let nodejs = NodeJS::new().build(context).await?;
        let pnpm = Pnpm::new().build(context).await?;
        let protoc = Protoc::new().build(context).await?;
        let protoc_gen_go = ProtocGenGo::new().build(context).await?;
        let protoc_gen_go_grpc = ProtocGenGoGrpc::new().build(context).await?;
        let staticcheck = Staticcheck::new().build(context).await?;

        artifact::ProjectEnvironment::new("vorpal-shell", SYSTEMS.to_vec())
            .with_artifacts(vec![
                bun,
                crane,
                go,
                goimports,
                gopls,
                grpcurl,
                nodejs,
                pnpm,
                protoc,
                protoc_gen_go,
                protoc_gen_go_grpc,
                staticcheck,
            ])
            .with_environments(vec![
                "CGO_ENABLED=0".to_string(),
                format!("GOARCH={}", get_goarch(context.get_system())?),
                format!("GOOS={}", get_goos(context.get_system())?),
            ])
            .with_secrets(vec![])
            .build(context)
            .await
    }
}

File: sdk/go/cmd/vorpal/artifact/vorpal_shell.go
package artifact

import (
	"fmt"

	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/artifact/language"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func BuildVorpalShell(context *config.ConfigContext) (*string, error) {
	contextTarget := context.GetTarget()

	bun, err := artifact.Bun(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get bun: %w", err)
	}

	crane, err := artifact.Crane(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get crane: %w", err)
	}

	gobin, err := artifact.GoBin(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get go: %w", err)
	}

	goimports, err := artifact.Goimports(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get goimports: %w", err)
	}

	gopls, err := artifact.Gopls(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get gopls: %w", err)
	}

	grpcurl, err := artifact.Grpcurl(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get grpcurl: %w", err)
	}

	nodejs, err := artifact.NodeJS(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get nodejs: %w", err)
	}

	pnpm, err := artifact.Pnpm(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get pnpm: %w", err)
	}

	protoc, err := artifact.Protoc(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get protoc: %w", err)
	}

	protocGenGo, err := artifact.ProtocGenGo(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get protoc-gen-go: %w", err)
	}

	protocGenGoGRPC, err := artifact.ProtocGenGoGRPC(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get protoc-gen-go-grpc: %w", err)
	}

	staticcheck, err := artifact.Staticcheck(context)
	if err != nil {
		return nil, fmt.Errorf("failed to get staticcheck: %w", err)
	}

	goarch, err := language.GetGOARCH(contextTarget)
	if err != nil {
		return nil, fmt.Errorf("failed to get GOARCH for target %s: %w", contextTarget, err)
	}

	goos, err := language.GetGOOS(contextTarget)
	if err != nil {
		return nil, fmt.Errorf("failed to get GOOS for target %s: %w", contextTarget, err)
	}

	return artifact.
		NewProjectEnvironment("vorpal-shell", SYSTEMS).
		WithArtifacts([]*string{
			bun,
			crane,
			gobin,
			goimports,
			gopls,
			grpcurl,
			nodejs,
			pnpm,
			protoc,
			protocGenGo,
			protocGenGoGRPC,
			staticcheck,
		}).
		WithEnvironments([]string{
			"CGO_ENABLED=0",
			fmt.Sprintf("GOARCH=%s", *goarch),
			fmt.Sprintf("GOOS=%s", *goos),
		}).
		Build(context)
}
